### ISSUE TITLE
lv2 build: confusion about parameter identities

### ISSUE THREAD

**Issue by jdeut** [surge-synthesizer/surge#4728]
**Bug Description:**

Some properties are  bounded unidirectionally. For example whenever i change the cutoff frequency of the second filter the controls of the filter balance also changes. The same holds for FM Depth and Scene Pitch for example.

**Surge Version**

Version:  Surge XT 1.0.main.bb6bf260
Build:  2021-07-05 @ 23:17:03 on 'johannes/local' with 'GNU-10.2.1' using JUCE 60008
System:  Linux 64-bit LV2 on  Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz

**Reproduction Steps:**

```
jalv.qt5 "https://surge-synthesizer.github.io/lv2/surge-xt"
```
and then see video

**Expected Behavior:**
surge controls are expected to be mutually indepent

**Screenshots:**

![out](https://user-images.githubusercontent.com/1739296/124671039-279d5000-deb5-11eb-9082-922aa5f91206.gif)

**Comment by baconpaul** [comment]
Hmm
I think I need to draft @KottV to help here since we don’t see it in vst3 au and standalone. Sounds like lv2 is reporting a change on a param offset from the changed one

**Comment by KottV** [comment]
confirmed, but I haven't idea yet

**Comment by baconpaul** [comment]
Ok some travel this week but am around - let me know how I can help

**Comment by mynameismuhl** [comment]
Hi @KottV, here is what I see with the LV2 in Reaper.
There is a reported offset of -4 for all parameters.
If I move macro C4 (in Surge) then last touched param (in Reaper) says C0
with "fx 1 return" I get C4 etc.

If I change line 472 in juce_LV2_Wrapper.cpp to
`controlPortOffset += 5;`
the problem seems to go away and "Last Touched" now states the correct parameter. 
That is how far I got.

To be able to compile the LV2
- with clang I had to comment out line 1387 in JUCEUtils.cmake as you suggested in #4488
  and add missing overrides in juce_LV2_Wrapper.cpp clang errors about.
- with gcc I have to set `-DCMAKE_CXX_FLAGS="-Wno-error=stringop-overflow="`.

Btw. is there a way to prevent CMake from installing 2440 presets everytime I rebuild?

**Comment by baconpaul** [comment]
We should just quadruple check that the VST3 doesn't exhibit this problem in your environment. I have only tested it mac vst3 reaper and logic mac au (but I think @mkruselj has also tested parameter work on windows with xt vst3).

**Comment by mynameismuhl** [comment]
If by -your environment- you mean Linux build with regular/non-alternate Juce, then no
this issue is not happening with VST3 in Ubuntu or Mint.

**Comment by baconpaul** [comment]
yup that's what I meant. exactly and appreciate it. Just want to make sure we are comparing apples and apples for @KottV (or I guess ubuntus and ubunutus :) )

**Comment by mynameismuhl** [comment]
oops, quadruple checked now and the GCC warning/error is also happening for VST3/Standalone (Release only).
I'll put that in a new issue. The rest of the above mentioned is LV2 only.

**Comment by KottV** [comment]
@falkTX 
do you have any thoughts?

>If I change line 472 in juce_LV2_Wrapper.cpp to
controlPortOffset += 5;
the problem seems to go away and "Last Touched" now states the correct parameter.
That is how far I got.

**Comment by falkTX** [comment]
There must be something missing on the wrapper, because the offset is calculated based on plugin macros (the offset is different depending on the build).

there is always 1 offset due to "freewheel" port (for realtime/offline processing info), the other 4 must be 2 for the audio and midi ports.

We need to find out why this offset is not correct then.

**Comment by falkTX** [comment]
Perhaps `getTotalNumOutputChannels` returns an incorrect value?

**Comment by falkTX** [comment]
Here is a request, please print after https://github.com/DISTRHO/juce/blob/main/modules/juce_audio_plugin_client/LV2/juce_LV2_Wrapper.cpp#L878
`findMaxTotalChannels (filter, numInChans, numOutChans);` the `numInChans` and `numOutChans`.

Then at https://github.com/DISTRHO/juce/blob/main/modules/juce_audio_plugin_client/LV2/juce_LV2_Wrapper.cpp#L476 print the `getTotalNumInputChannels` and `getTotalNumOutputChannels`

My guess is that these are different.

Also, I would be grateful if someone could paste the (main plugin file) generated ttl.
Thanks!

**Comment by baconpaul** [comment]
Surge has 2 inputs and 2-6 outputs si that is 4 indeed

**Comment by mynameismuhl** [comment]
For L878 I get:
```
numInChans 2
numOutChans 6
```
Trying to print
`getTotalNumInputChannels` and `getTotalNumOutputChannels`
gives me "was not declared in this scope" error.
(main plugin file) generated ttl, I assume you mean this one:
[SurgeXT.ttl.zip](https://github.com/surge-synthesizer/surge/files/6855822/SurgeXT.ttl.zip)

**Comment by falkTX** [comment]
> Trying to print
> `getTotalNumInputChannels` and `getTotalNumOutputChannels`
> gives me "was not declared in this scope" error.

err, I was shorting the line but you have the full thing on the code just a bit above the supposed print. `filter->getTotal...()`

thanks for the ttl, and the numChan print.

There must be something Surge is doing that I didn't anticipate for on the LV2 wrapper.
Is it using meta-parameters? does it have a dedicated bypass parameter?

The offset by 5 is very weird.
My only guess is 4 (6-2 from audio outputs) and 1 more from bypass.. but very wild guess..

**Comment by mynameismuhl** [comment]
ah yes
so here is the other print
```
TotalNumInputChannels 2
TotalNumOutputChannels 2
```

> The offset by 5 is very weird.

I had changed that offset to +=5 because it was +=1 before.

**Comment by falkTX** [comment]
oh, alright, so it is just missing 4 ports.
then I have a pretty good idea what is going on and how to fix it.

**Comment by falkTX** [comment]
Should be fixed in https://github.com/DISTRHO/JUCE/commit/1e28e8c1964faa7ee7478177439bddffc904a2a8

**Comment by mynameismuhl** [comment]
Yes, if I copy @falkTX 's changes into @KottV 's LV2 branch the issue seems to be fixed.
(except for the clang errors mentioned above ofc.).

**Comment by baconpaul** [comment]
Surge has two shapes of output bus, either stereo or 3 stereo. In reaper vst3 that shows up as a 1 or 3 output synth. In logic you can instantiste it either as a stereo or 3 stereo output synth. If you run vst3 or au validator on it you will see the optional output routings

sounds like lv2 is making ine choice one place and another choice another?

**Comment by falkTX** [comment]
LV2 has no concept of optional buses.
It could be easily added as an extension, but no one bothered with that yet.
So for the JUCE LV2 wrapper, I just enable all possible audio ports.

**Comment by baconpaul** [comment]
Ahh well then. That might be an undesirable choice for surge. Basically we output scenes onto each of the aux busses but if your daw plays then all you will get nasty problems if you don’t route properly (like signal without effects).

**Comment by falkTX** [comment]
hmm maybe best to disable the extra buses on LV2 build then.
How much is this feature used?

From what I have seen from users, though I know my experience is very biased, most of them dont use extra buses at all. With sidechain for compressors and limiters being the exception.

**Comment by baconpaul** [comment]
Ugh I don’t know how to do that easily. The part of the code which configures buses is complied once so I can’t conditional compile on plug-in. Hold on…

**Comment by baconpaul** [comment]
https://github.com/surge-synthesizer/surge/blob/02161336b02a8c913c24865e48f16ceed28cc564/src/surge_synth_juce/SurgeSynthProcessor.cpp#L22

Right so I guess one needs to move that from a constructor argument to an if based on wrapper type inside the constructor body?

**Comment by baconpaul** [comment]
Oh but The it has to be a positive if in vst3 au standalone not a negative on lv2 since the wrapper type lv2 won’t be there in our mainline build

**Comment by falkTX** [comment]
> https://github.com/surge-synthesizer/surge/blob/02161336b02a8c913c24865e48f16ceed28cc564/src/surge_synth_juce/SurgeSynthProcessor.cpp#L22
> 
> Right so I guess one needs to move that from a constructor argument to an if based on wrapper type inside the constructor body?

Just put the constructor value as a return type of a function?
`AudioProcessor(getBestMatchingBusesProperties(/*something plugin wrapper type here*/))`

**Comment by baconpaul** [comment]
And it is a widely used feature by a very small set of users who route creatively but it is fine for lv2 to not have it

**Comment by falkTX** [comment]
yeah, if they want that specific feature they can use the VST3 variant.
sounds quite fair to me :)

**Comment by baconpaul** [comment]
Lemme take a few minutes and see if I can write that conditional code and have vst3 and au work still

**Comment by falkTX** [comment]
btw... I know this is not the place but since I got your attention just want to confirm something.
is the non-xt surge still going to be supported moving on, or is that straight abandoned once XT version is ready?

**Comment by baconpaul** [comment]
We don't plan on releasing a 1.9.1 that's right.
The code is a straight line single repo from 1.9.0 to XT is the 'next version'
and we really only support head

**Comment by baconpaul** [comment]
OK I think I see a solution but I don't have an LV2 env running right now.

Can someone try the branch above (`baconpaul/lv2-shape-4728`) which should force the LV2 to two outputs only? If that works I can merge it in a PR and close this issue.

**Comment by mynameismuhl** [comment]
compiled baconpaul/lv2-shape-4728 with lv2-porting-project/JUCE lv2-branch
(without falks above changes)
available outputs in reaper are now 2
and the params are reported correctly.

**Comment by baconpaul** [comment]
Great. I will merge it then and close this once ci is done.
Thanks all!

**Comment by mkruselj** [comment]
Gotta make sure to not show the Workflow->Activate individual scene outputs option in the menu for LV2 only then, too...

**Comment by baconpaul** [comment]
Why the default to false? Agree on menu hide. Will figure out the general way to do that

**Comment by KottV** [comment]
i merged falkTX's fix in my fork

**Comment by mkruselj** [comment]
> Why the default to false? Agree on menu hide. Will figure out the general way to do that

Well for LV2 it should default to false, no? Or rather - disregard what is stored for that option in case of LV2, always use just the two outputs.

**Comment by baconpaul** [comment]
Yeah this pr forces it to false for the lv2
